### PR TITLE
Direct Solvers device fix

### PR DESCRIPTION
**Pull request by psocratis**
Small changes mostly inside the UMFPack/MUMPS/ComplexMUMPS/SuperLU `Mult` methods to sync load and solution data on host. This addresses the cases where `-no-pa -d 'device'` is used along with the direct solvers above. 

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1890](https://github.com/mfem/mfem/pull/1890) | @psocratis | @tzanio | @YohannDudouit + @camierjs + @jandrej + @dylan-copeland | 11/30/20 | 12/15/20 | 12/24/20 | |

### CODE REVIEW

File: linalg/hypre.cpp [hunk 1]
```diff
@@ -2511,8 +2514,16 @@ void HypreSolver::Mult(const HypreParVector &b, HypreParVector &x) const
...
    }
+
+   if (!iterative_mode)
+   {
+      x = 0.0;
+   }
+
    if (!setup_called)
    {
+      b.HostRead();
```
Comment (dylan-copeland) [inline linalg/hypre.cpp:2525]: Can these two lines be moved to line 2523, just before `if (!setup_called)`, to eliminate the same lines currently on 2540-2541?

**Comment by dylan-copeland** [comment]
@psocratis This PR introduces a sample run in `ex3p` that uses `HyprePCG` and `HypreAMS`. Would it be hard to add sample runs or serial unit tests (with the flag `[CUDA]`, run by `cunit_tests`) that use the direct solvers?

File: linalg/superlu.cpp [hunk 2]
```diff
@@ -523,9 +523,10 @@ void SuperLUSolver::Mult( const Vector & x, Vector & y ) const
    // SuperLU overwrites x with y, so copy x to y and pass that to the solve
    // routine.
 
-   y = x;
+   const double *xptr = x.HostRead();
```
Comment (dylan-copeland) [inline linalg/superlu.cpp:526]: For consistency, there could be `xptr` and `yptr` or `xPtr` and `yPtr`.

**Comment by psocratis** [comment]
> @psocratis This PR introduces a sample run in `ex3p` that uses `HyprePCG` and `HypreAMS`. Would it be hard to add sample runs or serial unit tests (with the flag `[CUDA]`, run by `cunit_tests`) that use the direct solvers?

@dylan-copeland, take a look at the latest commit. Is that what you had in mind?

**Comment by dylan-copeland** [comment]
@psocratis The unit test looks good, and it passes for me on lassen, but instead of testing multiple systems (different dimensions and order), I think it would be sufficient to test just one system (e.g. order 3, 2D, ne=3?), maybe with a loop over the 4 solvers this PR has enabled. In other words, test multiple solvers rather than multiple systems.

**Review by dylan-copeland — APPROVED** [review]

File: tests/unit/linalg/test_direct_solvers.cpp [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+         delete fec;
+         delete mesh;
+      }
+   }
+}
+
+#endif
+
+#ifdef DIRECT_SOLVE_PARALLEL
+
```
Comment (dylan-copeland) [inline tests/unit/linalg/test_direct_solvers.cpp:171]: I think this needs `#ifdef MFEM_USE_MPI` as in the other parallel tests.

Comment (psocratis) [reply tests/unit/linalg/test_direct_solvers.cpp:171]: if `MFEM_USE_MUMPS=YES` or `MFEM_USE_SUPERLU=YES` then `MFEM_USE_MPI` is implied. So I think the additional check is not necessary

**Comment by camierjs** [comment]
@psocratis, do you still need the `tests/unit/linalg/test_umf.cpp` file?

**Comment by psocratis** [comment]
> @psocratis, do you still need the `tests/unit/linalg/test_umf.cpp` file?

Good catch, no I don't. I removed it. Thanks.

**Comment by camierjs** [comment]
LGTM

**Review by camierjs — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by tzanio** [comment]
Re-merged in `next` for testing ...

**Comment by tzanio** [comment]
Re-merged in `next` for testing ...

**Review by jandrej — APPROVED** [review]

**Review by tzanio — APPROVED** [review]
Thanks for fixing this @psocratis!

@camierjs, @jandrej and @dylan-copeland -- thank you for the review

**Pull request merged**